### PR TITLE
NMT Perceiver Encoder

### PR DESCRIPTION
**Pull request by michalivne**
This PR separates the learning framework (i.e., loss) from the architecture, and adds a new Perceiver encoder bottleneck architecture.

# Summary

A few major changes are introduced here:

1. Separating the learning framework (i.e., loss, or model type) from the architecture of the encoder / decoder.
2. The bottleneck encoder transformer can now be constructed using get_nemo_transformer method.
3. Supported  learning frameworks: 
    * NLL - Conditional cross entropy (the usual NMT loss)
    * VAE - Variational Auto-Encoder (https://arxiv.org/pdf/1312.6114.pdf)
    * MIM - Mutual Information Machine (https://arxiv.org/pdf/2003.02645.pdf)
4. Supported encoder architectures: 
    * seq2seq - the usual transformer encoder without a bottleneck
    * bridge - attention bridge bottleneck (https://arxiv.org/pdf/1703.03130.pdf)
    * perceiver (**new**) -  Perceiver bottleneck (https://arxiv.org/pdf/2103.03206.pdf)

The changes allow to easily select the architecture via configuration. Specifically, using a shared base YAML configuration and overriding flags during experimentation allows easy comparison of bottleneck and non-bottleneck architectures. Separating the loss from the architecture leads to a cleaner code which is easier to understand.

# YAML Configuration

The following configurations in YAML config allows to control the learning framework and architecture:

```
model:
  ...
  model_type: 'nll' # learning (i.e., loss) type: nll (i.e., cross-entropy/auto-encoder), mim, vae (see description above)
  min_logv: -6 # minimal allowed log variance for mim
  latent_size: -1 # dimension of latent (projected from hidden) -1 will take value of hidden size
  non_recon_warmup_batches: 200000 # warm-up steps for mim, and vae losses
  recon_per_token: true # when false reconstruction is computed per sample, not per token
...
  encoder:
    library: nemo
    ...
    arch: seq2seq # seq2seq, bridge, perceiver (see description above)
    hidden_steps: 32 # fixed number of hidden steps
    hidden_blocks: 1 # number of repeat blocks (see classes for description)
    hidden_init_method: default # see classes for available values

  decoder:
    library: nemo
    ...
    arch: seq2seq # currently only seq2seq is supported
```

Detailed description of config parameters:

* **model.encoder.arch=seq2seq**
    * *model.encoder.hidden_steps is ignored*
    * *model.encoder.hidden_blocks is ignored*
    * *model.encoder.hidden_init_method is ignored*
* **model.encoder.arch=bridge**
    * *model.encoder.hidden_steps:* input is projected to the specified fixed steps 
    * *model.encoder.hidden_blocks:* number of encoder blocks to repeat after attention bridge projection
    * *model.encoder.hidden_init_method:*
         *  enc_shared (default) - apply encoder to inputs, than attention bridge, followed by hidden_blocks number of the same encoder (pre and post encoders share parameters)
         * identity - apply attention bridge to inputs, followed by hidden_blocks number of the same encoder
         * enc - similar to enc_shared but the initial encoder has independent parameters
* **model.encoder.arch=perceiver**
    * *model.encoder.hidden_steps:* input is projected to the specified fixed steps 
    * *model.encoder.hidden_blocks:* number of cross-attention + self-attention blocks to repeat after initialization block (all self-attention and cross-attention share parameters)
    * *model.encoder.hidden_init_method:*
         * params (default) - hidden state is initialized with learned parameters followed by cross-attention with independent parameters
         * bridge - hidden state is initialized with an attention bridge

# Usage Example

**NOTE:** ```enc_dec_nmt-bottleneck.py``` must be used. ```enc_dec_nmt.py``` does not support the required configuration to run properly and will likely fail.

```
python -- enc_dec_nmt-bottleneck.py \
      --config-path=conf \
      --config-name=aayn_bottleneck \
      ...
      model.model_type=nll \
      model.non_recon_warmup_batches=7500 \
      model.encoder.arch=perceiver \
      model.encoder.hidden_steps=32 \
      model.encoder.hidden_blocks=2 \
      model.encoder.hidden_init_method=params \
      ...
```

# Additional Info

* Classes *TransformerBottleneckEncoderNM*, *TransformerBottleneckDecoderNM*, and corresponding config classes *NeMoTransformerBottleneckConfig*, *NeMoTransformerBottleneckEncoderConfig*, *NeMoTransformerBottleneckDecoderConfig* were added to ```nemo/collections/nlp/modules/common/transformer/transformer_bottleneck.py```
* Class *PerceiverEncoder* was added to ```nemo/collections/nlp/modules/common/transformer/perceiver_encoders.py```
* Class *BridgeEncoder* was added to ```nemo/collections/nlp/modules/common/transformer/bridge_encoders.py```
* Classes *TransformerBottleneckEncoderNM* and *TransformerEncoderNM* test for src/tgt ids in the tokenizer valid range.

### CODE REVIEW

File: nemo/collections/nlp/models/machine_translation/mt_enc_dec_bottleneck_model.py [hunk 1]
```diff
@@ -49,51 +62,38 @@ class MTBottleneckModel(MTEncDecModel):
...
+        self.att_bridge_inner_size: int = cfg.get("att_bridge_inner_size", 1024)
+
+        # TODO: add label smoothing support for per-sample reconstruction loss
         if not self.recon_per_token:
             loss_fn = NLLLoss(ignore_index=self.decoder_tokenizer.pad_id, reduction='none',)
             self.loss_fn = self.eval_loss_fn = loss_fn
 
-        if self.model_type not in ["seq2seq", "seq2seq-br", "seq2seq-mim", "seq2seq-vae"]:
+        if self.model_type not in ["recon_only", "mim", "vae"]:
             raise ValueError("Unknown model_type = {model_type}".format(model_type=self.model_type,))
```
Comment (ericharper) [inline nemo/collections/nlp/models/machine_translation/mt_enc_dec_bottleneck_model.py:85]: Just a nitpick, but you could use an f-string here:
`f"Unknown model type: {self.model_type}"`

File: nemo/collections/nlp/models/machine_translation/mt_enc_dec_bottleneck_model.py [hunk 2]
```diff
@@ -129,51 +129,31 @@ def list_available_models(cls) -> Optional[Dict[str, str]]:
...
-            z_mask = torch.ones(z.shape[0:2]).to(hidden_mask)
-
-        if return_ortho_loss:
-            return z, z_mean, z_logv, z_mask, ortho_loss
-        else:
-            return z, z_mean, z_logv, z_mask
+        return z, z_mean, z_logv
 
     @typecheck()
     def forward(self, src, src_mask, tgt, tgt_mask, labels, train=True, return_info=False):
```
Comment (ericharper) [inline nemo/collections/nlp/models/machine_translation/mt_enc_dec_bottleneck_model.py:159]: We have training_step. Why do we need train=True in the forward? Can we move that logic to training_step?

File: nemo/collections/nlp/modules/common/transformer/bridge_encoders.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
```
Comment (ericharper) [inline nemo/collections/nlp/modules/common/transformer/bridge_encoders.py:1]: 2021

File: nemo/collections/nlp/modules/common/transformer/perceiver_encoders.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
```
Comment (ericharper) [inline nemo/collections/nlp/modules/common/transformer/perceiver_encoders.py:1]: 2021

**Review by ericharper — CHANGES_REQUESTED** [review]
Overall really nice PR. Just a few minor comments. My main concern is having `train=True` flag in forward. Can we remove that and just have the necessary logic in `training_step`?

File: nemo/collections/nlp/models/machine_translation/mt_enc_dec_bottleneck_model.py (hunk 1 above)
Comment (michalivne) [reply nemo/collections/nlp/models/machine_translation/mt_enc_dec_bottleneck_model.py:85]: Done

File: nemo/collections/nlp/modules/common/transformer/bridge_encoders.py (hunk 3 above)
Comment (michalivne) [reply nemo/collections/nlp/modules/common/transformer/bridge_encoders.py:1]: Done

File: nemo/collections/nlp/modules/common/transformer/perceiver_encoders.py (hunk 4 above)
Comment (michalivne) [reply nemo/collections/nlp/modules/common/transformer/perceiver_encoders.py:1]: Done

File: nemo/collections/nlp/models/machine_translation/mt_enc_dec_bottleneck_model.py (hunk 2 above)
Comment (michalivne) [reply nemo/collections/nlp/models/machine_translation/mt_enc_dec_bottleneck_model.py]: Done. Added loss() method to compute the loss which takes the train=True/False flag.

File: nemo/collections/nlp/models/machine_translation/mt_enc_dec_bottleneck_model.py (hunk 1 above)
Comment (michalivne) [reply nemo/collections/nlp/models/machine_translation/mt_enc_dec_bottleneck_model.py]: Done.

**Review by ericharper — APPROVED** [review]
LGTM. Great PR! Thanks for all of the changes.

**Review by MaximumEntropy — APPROVED** [review]
LGTM. Thanks

File: examples/nlp/machine_translation/conf/aayn_bottleneck.yaml [hunk 5]
```diff
@@ -1,10 +1,20 @@
-# The bottleneck architecture supports two architectures:
-#   1) No bottleneck: model_type=seq2seq, the usual NMT mdoel.
-#   2) Fixed-size bottleneck: model_type in [seq2seq-br, seq2seq-mim, seq2seq-vae],
-#      where the output of the encoder is projected to a fixed number of steps.
-# The loss of seq2seq-br is reconstruction only (like seq2seq).
-# The losses of seq2seq-mim, seq2seq-vae are MIM (https://arxiv.org/pdf/2003.02645.pdf)
-# and VAE (https://arxiv.org/pdf/1312.6114.pdf) correspondingly.
+# The bottleneck architecture supports three learning framework (i.e., losses)
+# via model.model_type:
+#   1) recon_only - reconstruction only loss (e.g., auto-encoder)
```
[the remark this answers is not in the document]
Comment (michalivne) [reply examples/nlp/machine_translation/conf/aayn_bottleneck.yaml]: Makes sense to me. Changed it recon_only --> nll (per our discussion)
